Exercise 1: Finite-precision arithmetic

Part_1 A 64 bits word size is commonly referred to as double precision,We would like to know: what is the roundoff error, what is the largest number, and what is the smallest number that can be represented in the computer?

In [6]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

max     = 1.7976931348623157e+308
min     = 2.2250738585072014e-308
epsilon = 2.220446049250313e-16

Part_2   
0.125 = 125/1000 = 1/8 =1/2^3 a good binary 0.001
and 0.25 = 1/4 = 1/2^2 also can write in binary 0.01
0.125+ 0.25 = 0.001 + 0.010 = 0.011
1/4 + 1/8 = 3/8=0.375
But! 
0.1 = 1/10 =1/(2*5)its representation keeps going forever computer cannot store infinite many digits so it cuts it down.same as 0.2 
so 0.1 +0.2 = 0.300000004
and therefore 0.1+0.2==0.3 is found FALSE.


Part_3
No we wont use == as it can come out false even though it isnt and it comes out false dure to a tiny difference. we can use math.isclose()

In [7]:
import math

a = 0.1 + 0.2
b = 0.3

print(a == b)
print(math.isclose(a, b))

False
True
